core/src/main/java/io/questdb/griffin/engine/functions/json/JsonPathFunctionFactory.java

```diff
@@ -0,0 +1,175 @@
+        final Function defaultValueFn = args.getQuiet(4);
+        final DirectUtf8Sink pointer = SupportingState.varcharConstantToJsonPointer(path);
+        final int maxSize = configuration.getStrFunctionMaxBufferLength();
+        final boolean strict = onError == JsonPathFunc.FAIL_ON_ERROR;
+        final JsonPathFunc fn = buildFunction(json, path, targetType, pointer, maxSize, strict);
+        applyDefault(fn, targetType, onError, defaultValueFn);
+        return fn;
+    }
+
+    private static @NotNull JsonPathFunc buildFunction(
+            Function json,
+            Function path,
+            int targetType,
+            DirectUtf8Sink pointer,
+            int maxSize,
+            boolean strict) {
+        switch (targetType) {
+            case ColumnType.LONG:
+            case ColumnType.DOUBLE:
+            case ColumnType.FLOAT:
+            case ColumnType.BOOLEAN:
+            case ColumnType.SHORT:
+            case ColumnType.INT:
+                return new JsonConstPathPrimitiveFunc(targetType, json, path, pointer, strict);
+            case ColumnType.VARCHAR:
+                return new JsonPathVarcharFunc(json, path, pointer, maxSize, false);
+            default:
+                throw new UnsupportedOperationException("nyi");  // TODO: complete remaining types
+        }
+    }
+
+    private static void applyDefault(JsonPathFunc fn, int targetType, int onError, Function defaultValueFn) {
+        if ((defaultValueFn == null) || (onError == JsonPathFunc.FAIL_ON_ERROR)) {
+            return;
+        }
+        if (!defaultValueFn.isConstant()) {
+            throw CairoException.nonCritical().put("default value must be constant");
+        }
+        switch (targetType) {
+            case ColumnType.LONG:
+                fn.setDefaultLong(defaultValueFn.getLong(null));
+                break;
+            case ColumnType.DOUBLE:
+                fn.setDefaultDouble(defaultValueFn.getDouble(null));
+                break;
+            case ColumnType.BOOLEAN:
+                fn.setDefaultBool(defaultValueFn.getBool(null));
+                break;
+            case ColumnType.SYMBOL:
+                fn.setDefaultSymbol(defaultValueFn.getSymbol(null));
+                break;
+            case ColumnType.SHORT:
+                fn.setDefaultShort(defaultValueFn.getShort(null));
+                break;
+            case ColumnType.INT:
+                fn.setDefaultInt(defaultValueFn.getInt(null));
+                break;
+            case ColumnType.FLOAT:
+                fn.setDefaultFloat(defaultValueFn.getFloat(null));
+                break;
+            case ColumnType.VARCHAR:
+                fn.setDefaultVarchar(defaultValueFn.getVarcharA(null));
+                break;
+            default:
+                // TODO: Better error messaging to use the name of the type, e.g. "DATE" instead of "7".
+                throw CairoException.nonCritical().put("unsupported target type: ").put(targetType);
```

Comment:
this is called from a factory, it should be SqlException and must include position of invalid argument

Reply:
Done.